### PR TITLE
chore(deps): bump revm to 6.0

### PR DESCRIPTION
**Pull request by mattsse**

### CODE REVIEW

File: crates/evm/core/src/utils.rs [hunk 1]
```diff
@@ -5,48 +5,49 @@ use eyre::ContextCompat;
...
+        SuccessReason::Return => InstructionResult::Return,
+        SuccessReason::Stop => InstructionResult::Stop,
+        SuccessReason::SelfDestruct => InstructionResult::SelfDestruct,
     }
 }
 
 /// Small helper function to convert a Halt into an InstructionResult
 #[inline]
-pub fn halt_to_instruction_result(halt: Halt) -> InstructionResult {
+pub fn halt_to_instruction_result(halt: HaltReason) -> InstructionResult {
```
Comment (DaniPopes) [inline crates/evm/core/src/utils.rs:29]: `HaltReason` is `Into<InstructionResult>`: <https://docs.rs/revm/latest/revm/interpreter/primitives/enum.HaltReason.html#impl-From%3CHaltReason%3E-for-InstructionResult>

File: crates/evm/core/src/utils.rs [hunk 2]
```diff
@@ -5,48 +5,49 @@ use eyre::ContextCompat;
...
 pub use crate::ic::*;
 
 /// Small helper function to convert an Eval into an InstructionResult
 #[inline]
-pub fn eval_to_instruction_result(eval: Eval) -> InstructionResult {
-    match eval {
-        Eval::Return => InstructionResult::Return,
-        Eval::Stop => InstructionResult::Stop,
-        Eval::SelfDestruct => InstructionResult::SelfDestruct,
+pub fn eval_to_instruction_result(reason: SuccessReason) -> InstructionResult {
```
Comment (DaniPopes) [inline crates/evm/core/src/utils.rs:19]: `SuccessReason` is `Into<InstructionResult>`: <https://docs.rs/revm/latest/revm/interpreter/primitives/enum.SuccessReason.html#impl-From%3CSuccessReason%3E-for-InstructionResult>

File: crates/chisel/src/runner.rs [hunk 3]
```diff
@@ -151,9 +151,7 @@ impl ChiselRunner {
                 self.executor.env.tx.gas_limit = mid_gas_limit;
                 let res = self.executor.call_raw(from, to, calldata.clone(), value)?;
                 match res.exit_reason {
-                    InstructionResult::Revert |
-                    InstructionResult::OutOfGas |
-                    InstructionResult::OutOfFund => {
+                    InstructionResult::Revert | InstructionResult::OutOfGas => {
```
Comment (DaniPopes) [inline crates/chisel/src/runner.rs:154]: `OutOfFunds`?

File: crates/evm/core/src/backend/fuzz.rs [hunk 4]
```diff
@@ -50,18 +51,25 @@ impl<'a> FuzzBackendWrapper<'a> {
...
+        let mut evm = revm::Evm::builder()
+            .with_db(self)
+            .with_external_context(inspector)
+            .with_env_with_handler_cfg(env)
+            .append_handler_register(inspector_handle_register)
+            .build();
+
+        match evm.transact() {
+            Ok(res) => Ok(res),
+            Err(err) => eyre::bail!("backend: failed while inspecting: {err}"),
```
Comment (DaniPopes) [inline crates/evm/core/src/backend/fuzz.rs:72]: `wrap_err("")`?

File: crates/evm/coverage/src/inspector.rs [hunk 5]
```diff
@@ -10,18 +10,16 @@ pub struct CoverageCollector {
...
     #[inline]
-    fn initialize_interp(&mut self, interpreter: &mut Interpreter<'_>, _: &mut EVMData<'_, DB>) {
-        let hash = interpreter.contract.hash;
+    fn initialize_interp(&mut self, interp: &mut Interpreter, context: &mut EvmContext<DB>) {
+        let hash = interp.contract.hash;
         self.maps.entry(hash).or_insert_with(|| {
-            HitMap::new(Bytes::copy_from_slice(
-                interpreter.contract.bytecode.original_bytecode_slice(),
-            ))
+            HitMap::new(Bytes::copy_from_slice(interp.contract.bytecode.original_bytecode_slice()))
```
Comment (DaniPopes) [inline crates/evm/coverage/src/inspector.rs:16]: [`original_bytecode`](https://docs.rs/revm/latest/revm/interpreter/struct.BytecodeLocked.html#method.original_bytecode)

File: crates/evm/evm/src/inspectors/chisel_state.rs [hunk 6]
```diff
@@ -22,12 +23,12 @@ impl ChiselState {
 
 impl<DB: Database> Inspector<DB> for ChiselState {
     #[inline]
-    fn step_end(&mut self, interp: &mut Interpreter<'_>, _: &mut revm::EVMData<'_, DB>) {
+    fn step(&mut self, interp: &mut Interpreter, context: &mut EvmContext<DB>) {
         // If we are at the final pc of the REPL contract execution, set the state.
         // Subtraction can't overflow because `pc` is always at least 1 in `step_end`.
         if self.final_pc == interp.program_counter() - 1 {
```
Comment (DaniPopes) [inline crates/evm/evm/src/inspectors/chisel_state.rs:29]: If this should be `step`, then this comment is not true anymore

File: crates/anvil/src/eth/backend/mem/inspector.rs [hunk 7]
```diff
@@ -48,94 +52,89 @@ impl Inspector {
...
-        data: &mut EVMData<'_, DB>,
-        call: &mut CallInputs,
-    ) -> (InstructionResult, Gas, Bytes) {
+        context: &mut EvmContext<DB>,
+        inputs: &mut CallInputs,
+        return_memory_offset: Range<usize>,
+    ) -> Option<CallOutcome> {
         call_inspectors!([&mut self.tracer, Some(&mut self.log_collector)], |inspector| {
-            inspector.call(data, call);
+            if let Some(outcome) = inspector.call(context, inputs, return_memory_offset) {
```
Comment (DaniPopes) [inline crates/anvil/src/eth/backend/mem/inspector.rs:90]: Is this right? If the tracer returns Some, log collector doesn't run

File: crates/evm/core/src/backend/mod.rs [hunk 8]
```diff
@@ -767,17 +768,24 @@ impl Backend {
...
+            .with_external_context(inspector)
+            .with_env_with_handler_cfg(env)
+            .append_handler_register(inspector_handle_register)
+            .build();
 
-        match revm::evm_inner::<Self>(env, self, Some(&mut inspector)).transact() {
+        match evm.transact() {
             Ok(res) => Ok(res),
-            Err(e) => eyre::bail!("backend: failed while inspecting: {e}"),
+            Err(err) => eyre::bail!("backend: failed while inspecting: {err}"),
```
Comment (DaniPopes) [inline crates/evm/core/src/backend/mod.rs:788]: wrap_err

File: crates/evm/core/src/backend/mod.rs [hunk 9]
```diff
@@ -1870,26 +1880,28 @@ fn update_env_block(env: &mut Env, fork_block: U64, block: &Block) {
...
-
-        match evm.inspect(inspector) {
+        let mut evm = revm::Evm::builder()
+            .with_db(db)
+            .with_external_context(inspector)
+            .with_env_with_handler_cfg(env)
+            .append_handler_register(inspector_handle_register)
+            .build();
+
+        match evm.transact() {
```
Comment (DaniPopes) [inline crates/evm/core/src/backend/mod.rs:1904]: wrap_err

File: crates/evm/core/src/opts.rs [hunk 10]
```diff
@@ -102,7 +102,6 @@ impl EvmOpts {
     pub fn local_evm_env(&self) -> revm::primitives::Env {
         let mut cfg = CfgEnv::default();
         cfg.chain_id = self.env.chain_id.unwrap_or(foundry_common::DEV_CHAIN_ID);
-        cfg.spec_id = SpecId::MERGE;
```
Comment (DaniPopes) [inline crates/evm/core/src/opts.rs:105]: is this right? Don't we override it on purpose?

File: crates/evm/core/src/utils.rs [hunk 11]
```diff
@@ -5,51 +5,15 @@ use eyre::ContextCompat;
 use foundry_config::NamedChain;
 use revm::{
     interpreter::InstructionResult,
-    primitives::{Eval, Halt, SpecId, TransactTo},
+    primitives::{SpecId, TransactTo},
 };
 
 pub use foundry_compilers::utils::RuntimeOrHandle;
 pub use revm::primitives::State as StateChangeset;
+use revm::primitives::{HaltReason, SuccessReason};
```
Comment (DaniPopes) [inline crates/evm/core/src/utils.rs:13]: move up

File: crates/evm/core/src/opts.rs (hunk 10 above)
Comment (mattsse) [reply crates/evm/core/src/opts.rs:105]: this no longer exists unfortunatelly...

File: crates/anvil/src/eth/backend/mem/mod.rs [hunk 12]
```diff
@@ -788,24 +788,29 @@ impl Backend {
...
-            Err(e) => return Err(e.into()),
+        let result_and_state = {
+            let mut evm = revm::Evm::builder()
+                .with_ref_db(&*db)
+                .with_external_context(&mut inspector)
+                .with_env_with_handler_cfg(env)
+                .append_handler_register(inspector_handle_register)
+                .build();
+
+            match evm.transact() {
```
Comment (DaniPopes) [inline crates/anvil/src/eth/backend/mem/mod.rs:799]: this looks like it can just be `?`

File: crates/anvil/src/eth/backend/mem/mod.rs [hunk 13]
```diff
@@ -1146,23 +1160,31 @@ impl Backend {
...
+            let result_and_state = {
+                let env = self.build_call_env(request, fee_details, block);
+                let mut evm = revm::Evm::builder()
+                    .with_ref_db(state)
+                    .with_external_context(&mut inspector)
+                    .with_env_with_handler_cfg(env)
+                    .append_handler_register(inspector_handle_register)
+                    .build();
+
+                    match evm.transact() {
```
Comment (DaniPopes) [inline crates/anvil/src/eth/backend/mem/mod.rs:1173]: use `?`

File: Cargo.toml [hunk 14]
```diff
@@ -154,22 +154,22 @@ ethers-middleware = { version = "2.0", default-features = false }
...
+alloy-providers = { git = "https://github.com/alloy-rs/alloy", rev = "785c667" }
+alloy-pubsub = { git = "https://github.com/alloy-rs/alloy", rev = "785c667" }
+alloy-rpc-client = { git = "https://github.com/alloy-rs/alloy", rev = "785c667" }
+alloy-rpc-trace-types = { git = "https://github.com/alloy-rs/alloy", rev = "785c667" }
+alloy-rpc-types = { git = "https://github.com/alloy-rs/alloy", rev = "785c667" }
+alloy-signer = { git = "https://github.com/alloy-rs/alloy", rev = "785c667" }
+alloy-transport = { git = "https://github.com/alloy-rs/alloy", rev = "785c667" }
+alloy-transport-http = { git = "https://github.com/alloy-rs/alloy", rev = "785c667" }
+alloy-transport-ipc = { git = "https://github.com/alloy-rs/alloy", rev = "785c667" }
+alloy-transport-ws = { git = "https://github.com/alloy-rs/alloy", rev = "785c667" }
```
Comment (DaniPopes) [inline Cargo.toml:172]: this is necessary for evm-inspectors right?

File: crates/cheatcodes/src/inspector.rs [hunk 15]
```diff
@@ -1412,12 +1481,14 @@ fn disallowed_mem_write(
...
+fn mstore_revert_string(interpreter: &mut Interpreter, bytes: &[u8]) {
     let starting_offset = interpreter.shared_memory.len();
     interpreter.shared_memory.resize(starting_offset + bytes.len());
     interpreter.shared_memory.set_data(starting_offset, 0, bytes.len(), bytes);
-    interpreter.return_offset = starting_offset;
-    interpreter.return_len = interpreter.shared_memory.len() - starting_offset
+
+    // TODO(mattsse): what happened to this??
+    // interpreter.return_offset = starting_offset;
+    // interpreter.return_len = interpreter.shared_memory.len() - starting_offset
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:1491]: This is now in `CallOutcome`, I think we want to pass `&mut CallOutcome` to this function

File: crates/evm/core/src/backend/mod.rs [hunk 16]
```diff
@@ -546,7 +549,7 @@ impl Backend {
     /// Sets the current spec id
     pub fn set_spec_id(&mut self, spec_id: SpecId) -> &mut Self {
         trace!("setting precompile id");
```
Comment (DaniPopes) [inline crates/evm/core/src/backend/mod.rs:551]:
```suggestion
        trace!(?spec_id, "setting spec ID");
```

File: crates/evm/evm/src/inspectors/chisel_state.rs (hunk 6 above)
Comment (DaniPopes) [reply crates/evm/evm/src/inspectors/chisel_state.rs:29]: Bump

File: crates/evm/evm/src/inspectors/stack.rs [hunk 17]
```diff
@@ -571,22 +578,15 @@ impl<DB: DatabaseExt + DatabaseCommit> Inspector<DB> for InspectorStack {
                 &mut self.printer
             ],
             |inspector| {
-                inspector.initialize_interp(interpreter, data);
-
-                // Allow inspectors to exit early
-                if interpreter.instruction_result != res {
```
Comment (DaniPopes) [inline crates/evm/evm/src/inspectors/stack.rs:577]: Why is this removed?

File: Cargo.toml (hunk 14 above)
Comment (mattsse) [reply Cargo.toml:172]: yeah, should we start publishing the rpc types?
this would make it a lot easier

File: crates/cheatcodes/src/inspector.rs (hunk 15 above)
Comment (DaniPopes) [reply crates/cheatcodes/src/inspector.rs:1491]: Actually we don't have that, this is during `step`

File: crates/evm/evm/src/inspectors/stack.rs (hunk 17 above)
Comment (mattsse) [reply crates/evm/evm/src/inspectors/stack.rs:577]: initialize is now infallible

File: Cargo.toml [hunk 18]
```diff
@@ -202,17 +204,11 @@ tracing = "0.1"
...
-# TODO: unpin on next revm release: https://github.com/bluealloy/revm/pull/870
-k256 = "=0.13.1"
+k256 = "0.13"
 
 axum = "0.6"
 hyper = "0.14"
 tower = "0.4"
 tower-http = "0.4"
 
 [patch.crates-io]
```
Comment (DaniPopes) [inline Cargo.toml:214]: :smile: 
```suggestion
```

**Review by DaniPopes — COMMENTED** [review]
Is this gtg?

File: Cargo.toml (hunk 14 above)
Comment (DaniPopes) [reply Cargo.toml:176]: We will with the rest of alloy, this is ok for now

**Comment by mattsse** [comment]
> Is this gtg?

I believe so

**Review by DaniPopes — APPROVED** [review]
